In [11]:
from PIL import Image
import pandas as pd
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np
import torchvision
from torchvision import transforms
from glob import glob
import os
import re

import shutil

# Internal data 누락된 정보 확인

In [7]:
internal_normal_path = '/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/original/internal_data/Control(normal)_img/*.jpg'
internal_abnormal_path = '/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/original/internal_data/Patient(abnormal)_img/*.jpg'

internal_nor_glob_path = glob(internal_normal_path)
internal_abnor_glob_path = glob(internal_abnormal_path)

print(len(internal_nor_glob_path))
print(len(internal_abnor_glob_path))

651
359


## renamed folder 생성

In [14]:
# os.makedirs('/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Control(normal)_img')
# os.makedirs('/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Patient(abnormal)_img')
renamed_normal_path = '/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Control(normal)_img'
renamed_abnormal_path = '/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Patient(abnormal)_img'

### internal_normal 이름 수정해서 파일 저장 

In [31]:
len_normal = len(internal_nor_glob_path)

bin_matrix = []

for i in range(len_normal):
    c = None
    d = None
    e = None
    f = None
    file = internal_nor_glob_path[i]
    #print(file)
    a= file.split('/')[-1]
    b= a.split('(')[-1]

    c = re.findall('05d+', b)
    d = re.findall('205d+', b)
    e = re.findall('305d+', b)
    f = re.findall('405d+', b)
    
#     print(c)
#     print(d)

    name_first = a.split('(')[0]
#한글과 영어이름을 띄어쓰기로 변환하고, 띄어쓰기를 다시 _로 변환
    korean = re.compile('[\u3131-\u3163\uac00-\ud7a3]+')
    name_first= re.sub(korean, '', name_first)
    name_first = re.sub('[a-zA-z]','',name_first).replace(' ', '_')
    name_first = re.sub(",", "", name_first)
    
# 마지막 글자가 다르거나 없는 경우가 존재하기 때문에 elif사용.    
    if d:
        name_last=d[0]
#         print(name_last)
        
    elif e:
        name_last=e[0]
        
    elif f:
        name_last=f[0]
        
    else:
        name_last=c[0]
        

    file_name = name_first+name_last+ '.jpg' 
    shutil.copy2(file, '/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Control(normal)_img/%s' % file_name)

    print(name_first+name_last)
    bin_matrix.append(file_name)

10524154__2020-03-09__05d
10527052__2018-03-17__05d
10527052__2019-02-16__05d
10527052__2021-09-25__05d
10536188__2019-05-27__05d
10536188__2021-02-24__205d
10536188__2021-02-24__05d
10567252__2020-01-06__05d
10586501__2019-04-08__205d
10586501__2021-07-21__05d
10614211__2016-03-24__05d
10619379__2020-03-24__05d
10626726__2020-12-17__05d
10626726__2022-01-04__205d
10626726__2022-01-04__05d
10639272__2021-06-22__205d
10639272__2021-06-22__05d
10781135__2020-02-26__05d
10794302__2020-06-11__205d
10794302__2020-06-11__05d
10794302__2021-06-04__05d
10796067__2020-09-15__05d
10801899__2018-12-01__05d
10822531__2020-01-06__05d
10822531__2021-01-09__05d
10822531__2022-01-08__05d
10836307__2019-12-16__05d
10836307__2020-12-02__05d
10864698__2018-03-05__05d
10898019__2019-07-09__05d
10904750__2019-01-30__05d
10904750__2020-01-06__05d
11074816__2019-06-01__05d
11074816__2020-08-29__05d
11079549__2015-07-04__205d
11079549__2015-07-04__05d
11111405__2019-12-06__05d
11113944__2018-03-22__05d
111139

### internal_abnormal 이름 수정해서 파일 저장 

In [35]:
len_abnormal = len(internal_abnor_glob_path)

bin_matrix = []

for i in range(len_abnormal):
    c = None
    d = None
    e = None
    f = None
    g = None
    h = None
    j = None
    k = None
    l = None
    
    file = internal_abnor_glob_path[i]
    #print(file)
    a= file.split('/')[-1]
    b= a.split('(')[-1]

    c = re.findall('05d+', b)
    d = re.findall('205d+', b)
    e = re.findall('305d+', b)
    f = re.findall('405d+', b)
    g = re.findall('505d+', b)
    h = re.findall('605d+', b)
    j = re.findall('705d+', b)
    k = re.findall('805d+', b)
    l = re.findall('905d+', b)
    
#     print(c)
#     print(d)

    name_first = a.split('(')[0]
#한글과 영어이름을 띄어쓰기로 변환하고, 띄어쓰기를 다시 _로 변환
    korean = re.compile('[\u3131-\u3163\uac00-\ud7a3]+')
    name_first= re.sub(korean, '', name_first)
    name_first = re.sub('[a-zA-z]','',name_first).replace(' ', '_')
    name_first = re.sub(",", "", name_first)
    
# 마지막 글자가 다르거나 없는 경우가 존재하기 때문에 elif사용.    

    if d:
        name_last=d[0]
#         print(name_last)
        
    elif e:
        name_last=e[0]
        
    elif f:
        name_last=f[0]
        
    elif g:
        name_last=g[0]
    
    elif h:
        name_last=h[0]
    
    elif j:
        name_last=j[0]
    
    elif k:
        name_last=k[0]
        
    elif l:
        name_last=l[0]
        
    else:
        name_last=c[0]

    file_name = name_first+name_last+ '.jpg' 
    shutil.copy2(file, '/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Patient(abnormal)_img/%s' % file_name)

    print(name_first+name_last)
    bin_matrix.append(file_name)

11725493__2017-04-11__205d
11725493__2017-04-11__305d
11725493__2017-04-11__05d
12608236__2020-06-19__205d
12608236__2020-06-19__305d
12608236__2020-06-19__05d
13131359__2018-05-02__205d
13131359__2018-05-02__05d
13227717__2018-08-13__205d
13227717__2018-08-13__305d
13227717__2018-08-13__05d
13334778__2020-04-01__205d
13334778__2020-04-01__305d
13334778__2020-04-01__05d
13562915__2017-03-20_205d
13562915__2017-03-20_05d
14104552__2020-10-12__05d
14232990__2017-10-17__05d
14410822__2016-05-02__205d
14410822__2016-05-02__405d
14410822__2016-05-02__05d
16856246__2016-10-24__05d
18363043__2018-09-15__205d
18363043__2018-09-15__305d
18363043__2018-09-15__405d
18363043__2018-09-15__05d
21029471__2020-07-10__05d
21032624__2018-07-12__205d
21032624__2018-07-12__305d
21032624__2018-07-12__05d
22520458__2018-08-28__205d
22520458__2018-08-28__05d
22734543__2015-10-06__05d
25274709__2020-12-11__205d
25274709__2020-12-11__305d
25274709__2020-12-11__05d
25583744__2015-12-07__205d
25583744__2015-12-0

## 누락된 파일 정리

In [80]:
internal_normal_info =  pd.read_excel('/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/Patients_info_v1.xlsx', sheet_name=1, engine='openpyxl')
internal_abnormal_info =  pd.read_excel('/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/Patients_info_v1.xlsx', sheet_name=0, engine='openpyxl')

re_in_normal_glob = glob('/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Control(normal)_img/*.jpg')
re_in_abnormal_glob = glob('/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Patient(abnormal)_img/*.jpg')

In [63]:
# normal & abnormal excel file 환자번호 추출
df_internal_normal = internal_normal_info['환자 번호']
df_internal_normal_list = df_internal_normal.values.tolist()
df_internal_normal_str_list = list(map(str, df_internal_normal_list))

df_internal_abnormal = internal_abnormal_info['환자번호']
df_internal_abnormal_list = df_internal_abnormal.values.tolist()
df_internal_abnormal_str_list = list(map(str, df_internal_abnormal_list))

In [64]:
# image folder에 있는 file들의 환자 번호를 추출하는데, 중복된 경우가 존재하기 때문에 중복제거

in_nor_img_list = []
for i in range(len(re_in_normal_glob)):
    nor_patient_num = re_in_normal_glob[i].split('/')[-1].split('_')[0]
#     print(patient_num)
    in_nor_img_list.append(nor_patient_num)
in_one_nor_img_list = list(set(in_nor_img_list))


in_abnor_img_list = []
for i in range(len(re_in_abnormal_glob)):
    abnor_patient_num = re_in_abnormal_glob[i].split('/')[-1].split('_')[0]
#     print(patient_num)
    in_abnor_img_list.append(abnor_patient_num)
in_one_abnor_img_list = list(set(in_abnor_img_list))


print(len(in_one_nor_img_list))
print(len(in_one_abnor_img_list))

401
129


- internal_data 상황

    - normal 
    1. 현재 이미지는 존재하지만 excel file에 존재하지 않는 경우 35개의 이미지 존재

    2. 현재 excel file에는 존재하지만 이미지에는 존재하지 않는 경우는 없다. 

    - abnormal 
    1. image folder에 존재하는 환자 수 : 129명

    2. excel file에 존재하는 환자 수 : 129명 -> 하지만 excel file에는 한 환자가 다른 날짜에 관찰한 경우가 2명의 환자 존재
        - 44205052(강효윤), 52721531(최민서)
----------------------------------------------------------------------------------------------------------------------
- summary    
    - 총 이미지를 기준으로 401명이 존재하지만 excel file에 35명이 존재하지 않아 366명이 존재하고, 반대의 경우는 존재하지 않는다.
        - internal_normal : 366명

    - 누락된 abnormal image X
        - internal_abnormal : 129명(하지만 excel file에는 한 환자가 다른 날짜에 관찰한 경우가 2명의 환자 존재)


||normal|abnormal|
|:---|---:|:---:|
|환자수|366명|129명|
|이미지수|612장|359장|

### normal check

In [59]:
# normal 
internal_img_no_csv_list = []
internal_img_csv_list = []
for i in in_one_nor_img_list:
#     print(i)
#     patient_num_img = abnormal_glob[i].split('/')[-1].split('_')[0]

    if i not in df_internal_normal_str_list:
        internal_img_no_csv_list.append(i)
#         print('파일이 존재하지 않습니다:',i)
    else:
        internal_img_csv_list.append(i)
#         print('파일이 존재:',i)

#     else:
#         print('있습니다!!')
        

print('image O, excel O 환자 수 : ', len(internal_img_csv_list))

print('image O, excel X 환자 수 : ', len(internal_img_no_csv_list))

image O, excel O 환자 수 :  366
image O, excel X 환자 수 :  35


In [58]:
internal_img_no_csv_list

['62203953',
 '62453664',
 '35214254',
 '11576622',
 '20689368',
 '54595118',
 '39239109',
 '62824088',
 '21531741',
 '12270662',
 '29312261',
 '18747254',
 '55447717',
 '36516674',
 '29990135',
 '62139425',
 '62629199',
 '28370909',
 '60898913',
 '12563023',
 '18345874',
 '10796067',
 '17414140',
 '56851535',
 '25790441',
 '12549421',
 '19006103',
 '11982342',
 '11877495',
 '10614211',
 '18709238',
 '54856837',
 '53072922',
 '52025567',
 '32329599']

In [60]:
# normal
internal_csv_no_img_list = []
internal_csv_img_list = []
for i in df_internal_normal_str_list:
#     print(i)
#     patient_num_img = abnormal_glob[i].split('/')[-1].split('_')[0]

    if i not in in_one_nor_img_list:
        internal_csv_no_img_list.append(i)
#         print('파일이 존재하지 않습니다:',i)
    else:
        internal_csv_img_list.append(i)
#         print('파일이 존재:',i)

#     else:
#         print('있습니다!!')
        

print('image O, excel O 환자 수 : ', len(internal_csv_img_list))

print('image O, excel X 환자 수 : ', len(internal_csv_no_img_list))

image O, excel O 환자 수 :  366
image O, excel X 환자 수 :  0


### abnormal check

In [66]:
# abnormal 
in_abnor_img_no_csv_list = []
in_abnor_img_csv_list = []
for i in in_one_abnor_img_list:


    if i not in df_internal_abnormal_str_list:
        in_abnor_img_no_csv_list.append(i)

    else:
        in_abnor_img_csv_list.append(i)

        

print('image O, excel O 환자 수 : ', len(in_abnor_img_csv_list))

print('image O, excel X 환자 수 : ', len(in_abnor_img_no_csv_list))

image O, excel O 환자 수 :  129
image O, excel X 환자 수 :  0


In [67]:
# abnormal 
in_abnor_csv_no_img_list = []
in_abnor_csv_img_list = []
for i in df_internal_abnormal_str_list:


    if i not in in_one_abnor_img_list:
        in_abnor_csv_no_img_list.append(i)

    else:
        in_abnor_csv_img_list.append(i)

        

print('image O, excel O 환자 수 : ', len(in_abnor_csv_img_list))

print('image O, excel X 환자 수 : ', len(in_abnor_csv_no_img_list))

image O, excel O 환자 수 :  133
image O, excel X 환자 수 :  0


In [68]:
print(len(df_internal_abnormal_str_list))
print(len(in_one_abnor_img_list))

133
129


## 이미지만 존재하는 하는 경우 제거

In [79]:
a = 0
for i in range(len(re_in_normal_glob)):
    
    patient_num = re_in_normal_glob[i].split('/')[-1].split('_')[0]
    
    if patient_num not in internal_img_csv_list:
        img_path = re_in_normal_glob[i]
        os.remove(img_path)
        a += 1

FileNotFoundError: [Errno 2] No such file or directory: '/home/crop2292/research_datasets/Upper_AoV_dysplasia/train_data/renamed/internal_data/Control(normal)_img/10614211__2016-03-24__05d.jpg'

In [83]:
nor_img_list = []
for i in range(len(re_in_normal_glob)):
    nor_patient_num = re_in_normal_glob[i].split('/')[-1].split('_')[0]
#     print(patient_num)
    nor_img_list.append(nor_patient_num)
in_one_nor_img_list = list(set(nor_img_list))



In [85]:
print(len(re_in_normal_glob))
print(len(in_one_nor_img_list))

612
366


## excel 존재 X, 이미지만 존재하는 경우 excel에서 제거

In [86]:
for i in len(df_internal_normal_list):
    patient_num = str(internal_normal_info['환자번호'][i])
    
    if patient_num not in internal_img_csv_list:
        df_drop = 
        

0      10605527
1      10684900
2      10719510
3      10747445
4      10912340
         ...   
361    63441215
362    62202936
363    10619379
364    11477174
365    23500673
Name: 환자 번호, Length: 366, dtype: int64

In [87]:
# normal & abnormal excel file 환자번호 추출
df_internal_normal = internal_normal_info['환자 번호']
df_internal_normal_list = df_internal_normal.values.tolist()
df_internal_normal_str_list = list(map(str, df_internal_normal_list))

df_internal_abnormal = internal_abnormal_info['환자번호']
df_internal_abnormal_list = df_internal_abnormal.values.tolist()
df_internal_abnormal_str_list = list(map(str, df_internal_abnormal_list))

0      10535370
1      11182414
2      11725493
3      12608236
4      13131359
         ...   
128    64737476
129    62559485
130    11058371
131    41324194
132    61042489
Name: 환자번호, Length: 133, dtype: int64

# external data 누락된 정보 확인